core/services/bulletprooftxmanager/eth_confirmer.go

```diff
@@ -216,28 +216,103 @@ func (ec *ethConfirmer) SetBroadcastBeforeBlockNum(blockNum int64) error {
 }
 
 func (ec *ethConfirmer) CheckForReceipts(ctx context.Context, blockNum int64) error {
-	batchSize := int(ec.config.EthRPCDefaultBatchSize())
-
 	attempts, err := ec.findEthTxAttemptsRequiringReceiptFetch()
+
 	if err != nil {
 		return errors.Wrap(err, "findEthTxAttemptsRequiringReceiptFetch failed")
 	}
 	if len(attempts) == 0 {
 		return nil
 	}
-	if batchSize == 0 {
-		batchSize = len(attempts)
-	}
 
 	logger.Debugw(fmt.Sprintf("EthConfirmer: fetching receipts for %v transaction attempts", len(attempts)), "blockNum", blockNum)
 
+	attemptsByAddress := make(map[gethCommon.Address][]models.EthTxAttempt)
+	for _, att := range attempts {
+		attemptsByAddress[att.EthTx.FromAddress] = append(attemptsByAddress[att.EthTx.FromAddress], att)
+	}
+
+	for from, attempts := range attemptsByAddress {
+		ctxInner, cancel := eth.DefaultQueryCtx(ctx)
+		latestBlockNonce, err := ec.getNonceForLatestBlock(ctxInner, from)
+		defer cancel()
+
+		if ctxInner.Err() != nil { // timeout
+			break
+		}
+		if err != nil {
+			return errors.Wrapf(err, "unable to fetch pending nonce for address: %v", from)
+		}
+
+		likelyConfirmed := separateLikelyConfirmedAttempts(from, attempts, latestBlockNonce)
+		likelyConfirmedCount := len(likelyConfirmed)
+		if likelyConfirmedCount != 0 {
+			likelyUnconfirmedCount := len(attempts) - likelyConfirmedCount
+
+			logger.Debugf("EthConfirmer: Fetching and saving %v likely confirmed receipts. Skipping checking the others (%v)",
+				likelyConfirmedCount, likelyUnconfirmedCount)
+
+			start := time.Now()
+			err = ec.fetchAndSaveReceipts(ctx, likelyConfirmed, blockNum)
+			if err != nil {
+				return errors.Wrapf(err, "unable to fetch and save receipts for likely confirmed txs, for address: %v", from)
+			}
+			logger.Debugf("EthConfirmer: Fetching and saving %v likely confirmed receipts took %v ms.",
+				likelyConfirmedCount, float64(time.Since(start).Milliseconds()))
+		}
+	}
+
+	if err := ec.markConfirmedMissingReceipt(); err != nil {
+		return errors.Wrap(err, "unable to mark eth_txes as 'confirmed_missing_receipt'")
+	}
+
+	if err := ec.markOldTxesMissingReceiptAsErrored(blockNum); err != nil {
+		return errors.Wrap(err, "unable to confirm buried unconfirmed eth_txes")
+	}
+	return nil
+}
+
+func separateLikelyConfirmedAttempts(from gethCommon.Address, attempts []models.EthTxAttempt, latestBlockNonce uint64) []models.EthTxAttempt {
+	firstAttemptNonce := "unknown"
+	lastAttemptNonce := "unknown"
+	if attempts[len(attempts)-1].EthTx.Nonce != nil {
```

Comment:
Can attempts be empty here

Reply:
It is reasonable to assume that the nonce is not nil and panic if so. EthConfirmer is not supposed to touch anything with a null nonce, so the nil check is unnecessary

Reply:
-> Can attempts be empty here 
No, there is a check in CheckForReceipts but still perhaps more clear to add another here.

-> EthConfirmer is not supposed to touch anything with a null nonce, so the nil check is unnecessary
Ah perfect then.